### DISCUSSION TITLE
Infinite recursion possible with defer?

### DISCUSSION THREAD

**Discussion by yaacovCR** [graphql/defer-stream-wg#29] [Q&A] [answered]
```graphql
      {
        ...Frag
      }

      fragment Frag on Type {
        field
        ...Frag @defer
      }
```

See: https://github.com/robrichard/graphql-spec/blame/incremental/spec/Section%206%20--%20Execution.md#L601-L602

I'm probable also missing something in terms of the use case. Why the exception for when defer is true?

**Comment by yaacovCR** [comment]
I guess we have a validation rule to prevent fragment cycles, so relaxing this within execution is an easy way to ensure that separately deferred named fragments are deferred/sent the actual requested number of times

**Reply by yaacovCR** [reply]
And I guess the question is whether this check should be removed for non validated documents with defer….

maybe we could allow sibling uses of defer with identical fragments but still disallow cycles by keeping a separate set of visited fragments that knows not of siblings

**Comment by robrichard** [comment, marked as answer]
Yes, the idea is to make sure all the deferred fragments get executed and non-deferred fragments don't get de-duped by deferred ones. I think the check needs to be modified so we also don't add deferred fragments to `visitedFragments`. I'm going to make this change now.

If you skip validation, it is possible for `execute` to recurse infinitely today without defer with a query like this:
```graphql
{
    hero {
        ...PersonFragment
    }
}

fragment PersonFragment on Person {
    friends {
         ...PersonFragment
    }
}
```
I'm not sure how important it is to guard against the additional scenario that defer introduces. It probably could be avoided with more complex logic.

**Reply by yaacovCR** [reply]
Sounds good to me, I think we expect operations to be validated in the general case so no huge deal
